src/Catel.Core/Catel.Core.Shared/Services/ObjectIdGenerator.cs

```diff
@@ -0,0 +1,72 @@
+﻿// --------------------------------------------------------------------------------------------------------------------
+// <copyright file="ObjectIdGenerator.cs" company="Catel development team">
+//   Copyright (c) 2008 - 2018 Catel development team. All rights reserved.
+// </copyright>
+// --------------------------------------------------------------------------------------------------------------------
+
+
+namespace Catel.Services
+{
+    using System.Collections.Generic;
+
+    /// <summary>
+    /// The ObjectIdGenerator class.
+    /// </summary>
+    /// <typeparam name="TObjectType">The object type</typeparam>
+    /// <typeparam name="TUniqueIdentifier">The unique identifier type</typeparam>
+    public abstract class ObjectIdGenerator<TObjectType, TUniqueIdentifier> : IObjectIdGenerator<TObjectType, TUniqueIdentifier>
+    {
+        private static Queue<TUniqueIdentifier> _releasedUniqueIdentifiers;
+
+        private static readonly object _syncObj = new object();
+
+        /// <inheritdoc />
+        public TUniqueIdentifier GetUniqueIdentifier()
+        {
+            return GenerateUniqueIdentifier();
+        }
+
+        /// <inheritdoc />
+        public TUniqueIdentifier GetUniqueIdentifier(bool reuse)
+        {
+            if (reuse)
+            {
+                lock (_syncObj)
+                {
+                    if (_releasedUniqueIdentifiers == null)
+                    {
+                        _releasedUniqueIdentifiers = new Queue<TUniqueIdentifier>();
```

Comment:
We don't even have to create it here, we only need to check whether it's != null. As long as we don't have released any identifiers, we don't need to check the collection. And the release will create the dictionary anyway.

Basically the check should be:

```
if (_releasedUniqueIdentifiers  != null && _releasedUniqueIdentifiers.Count > 0)
{
    return _releasedUniqueIdentifiers.Dequeue(); 
}
```

Reply:
Fixed

Reply:
Fixed